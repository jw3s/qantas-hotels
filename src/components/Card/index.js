import React, { PureComponent } from 'react';
import Spinner from 'react-loader-spinner';
import type { Element } from 'react';
import styled from 'styled-components';

const CardWrapper = styled.div`
  box-shadow: rgba(0, 0, 0, 0.08) 0px 4px 8px;
  display: flex;
  background: #fff;
  margin-bottom: 10px;
  color: #54545e;
  @media (max-width: 900px) {
    flex-direction: column;
  }
  overflow: hidden;
  position: relative;
`;

const CardImageWrapper = styled.div`
  overflow: hidden;
  flex: 0 0 25%;
  min-width: 250px;
  display: flex;
  @media (max-width: 900px) {
    max-height: 250px;
  }
  > div {
    margin: auto;
  }
`;
const CardImage = styled.img`
  width: 100%;
  height: 100%;
  transition: transform 200ms ease;
  :hover {
    transform: scale(1.1);
  }
  ${({ status }) => status === 'loading' && { display: 'none' }};
`;

const CardBody = styled.div`
  flex: 1;
  display: flex;
  padding: 20px;
  flex-wrap: wrap;
`;

type Props = {
  image?: string,
  children: Element<*> | Array<Element<*>>,
};
type State = {
  imageStatus: 'loading' | 'loaded' | 'failed',
};

class Card extends PureComponent<Props, State> {
  state = {
    imageStatus: 'loading',
  };

  static defaultProps = {
    image: '',
  };

  render() {
    const { imageStatus } = this.state;
    const { image, children, ...props } = this.props;
    return (
      <CardWrapper {...props}>
        {image && (
          <CardImageWrapper>
            {imageStatus === 'loading' && (
              <Spinner type="ThreeDots" color="#cd0000" height={30} width={30} />
            )}

            <CardImage
              status={imageStatus}
              onLoad={() => {
                this.setState({ imageStatus: 'loaded' });
              }}
              src={image}
            />
          </CardImageWrapper>
        )}
        <CardBody>{children}</CardBody>
      </CardWrapper>
    );
  }
}

export default Card;
