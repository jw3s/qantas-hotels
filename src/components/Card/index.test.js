import React from 'react';
import renderer from 'react-test-renderer';
import Card from '.';

import 'jest-styled-components';

describe('<Card />', () => {
  it('<Card /> should render tree', () => {
    const tree = renderer
      .create(
        <Card image="https://unsplash.it/700/500?random&mp;sig=1">
          <div>Card test</div>
        </Card>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('<Card /> should not render image if not provided', () => {
    const tree = renderer
      .create(
        <Card>
          <div>Card test</div>
        </Card>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
