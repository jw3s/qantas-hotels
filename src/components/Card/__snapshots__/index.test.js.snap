// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> <Card /> should not render image if not provided 1`] = `
.c0 {
  box-shadow: rgba(0,0,0,0.08) 0px 4px 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #fff;
  margin-bottom: 10px;
  color: #54545e;
  overflow: hidden;
  position: relative;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

@media (max-width:900px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  className="c0"
>
  
  <div
    className="c1"
  >
    <div>
      Card test
    </div>
  </div>
</div>
`;

exports[`<Card /> <Card /> should render tree 1`] = `
.c0 {
  box-shadow: rgba(0,0,0,0.08) 0px 4px 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #fff;
  margin-bottom: 10px;
  color: #54545e;
  overflow: hidden;
  position: relative;
}

.c1 {
  overflow: hidden;
  -webkit-flex: 0 0 25%;
  -ms-flex: 0 0 25%;
  flex: 0 0 25%;
  min-width: 250px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 > div {
  margin: auto;
}

.c2 {
  width: 100%;
  height: 100%;
  -webkit-transition: -webkit-transform 200ms ease;
  -webkit-transition: transform 200ms ease;
  transition: transform 200ms ease;
  display: none;
}

.c2:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

@media (max-width:900px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:900px) {
  .c1 {
    max-height: 250px;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div>
      <svg
        fill="#cd0000"
        height={30}
        viewBox="0 0 120 30"
        width={30}
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="15"
          cy="15"
          r="15"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="15"
            repeatCount="indefinite"
            to="15"
            values="15;9;15"
          />
          <animate
            attributeName="fillOpacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="1"
            repeatCount="indefinite"
            to="1"
            values="1;.5;1"
          />
        </circle>
        <circle
          attributeName="fillOpacity"
          cx="60"
          cy="15"
          from="1"
          r="9"
          to="0.3"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="9"
            repeatCount="indefinite"
            to="9"
            values="9;15;9"
          />
          <animate
            attributeName="fillOpacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="0.5"
            repeatCount="indefinite"
            to="0.5"
            values=".5;1;.5"
          />
        </circle>
        <circle
          cx="105"
          cy="15"
          r="15"
        >
          <animate
            attributeName="r"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="15"
            repeatCount="indefinite"
            to="15"
            values="15;9;15"
          />
          <animate
            attributeName="fillOpacity"
            begin="0s"
            calcMode="linear"
            dur="0.8s"
            from="1"
            repeatCount="indefinite"
            to="1"
            values="1;.5;1"
          />
        </circle>
      </svg>
    </div>
    <img
      className="c2"
      onLoad={[Function]}
      src="https://unsplash.it/700/500?random&mp;sig=1"
    />
  </div>
  <div
    className="c3"
  >
    <div>
      Card test
    </div>
  </div>
</div>
`;
